File: backend/app/services/google_auth_service.py
import secrets
from datetime import timedelta

from flask import current_app
from flask_jwt_extended import create_access_token, get_jwt
from google.auth.transport import requests as google_requests
from google.oauth2 import id_token

from app.constants import JAMAICA_PARISHES
from app.extensions import db
from app.models.user import User
from app.services.auth_service import hash_password, normalize_phone
from app.services.email_service import EmailServiceError, send_welcome_email
from app.services.shipping_id_service import generate_shipping_id
from app.services.token_service import access_token_for_user
from app.services.trn_service import normalize_trn
from app.services.warehouse_service import build_shipping_address

GOOGLE_SIGNUP_TOKEN_TYPE = "google_signup"
GOOGLE_SIGNUP_TOKEN_MINUTES = 15


class GoogleAuthError(ValueError):
    pass


def _google_client_id() -> str:
    client_id = (current_app.config.get("GOOGLE_CLIENT_ID") or "").strip()
    if not client_id:
        raise GoogleAuthError("Google sign-in is not configured")
    return client_id


def verify_google_credential(credential: str) -> dict:
    if not credential or not credential.strip():
        raise GoogleAuthError("Google credential is required")

    try:
        payload = id_token.verify_oauth2_token(
            credential.strip(),
            google_requests.Request(),
            _google_client_id(),
        )
    except ValueError as exc:
        raise GoogleAuthError("Invalid Google sign-in token") from exc

    if payload.get("iss") not in ("accounts.google.com", "https://accounts.google.com"):
        raise GoogleAuthError("Invalid Google token issuer")

    email = (payload.get("email") or "").strip().lower()
    if not email:
        raise GoogleAuthError("Google account email is required")
    if not payload.get("email_verified"):
        raise GoogleAuthError("Google account email is not verified")

    google_id = payload.get("sub")
    if not google_id:
        raise GoogleAuthError("Invalid Google account")

    first_name = (payload.get("given_name") or email.split("@")[0]).strip()
    last_name = (payload.get("family_name") or "").strip() or "Member"

    return {
        "google_id": str(google_id),
        "email": email,
        "first_name": first_name[:80],
        "last_name": last_name[:80],
    }


def create_google_signup_token(profile: dict) -> str:
    return create_access_token(
        identity=f"google:{profile['google_id']}",
        additional_claims={
            "type": GOOGLE_SIGNUP_TOKEN_TYPE,
            "google_id": profile["google_id"],
            "email": profile["email"],
            "first_name": profile["first_name"],
            "last_name": profile["last_name"],
        },
        expires_delta=timedelta(minutes=GOOGLE_SIGNUP_TOKEN_MINUTES),
    )


def _read_google_signup_claims() -> dict:
    claims = get_jwt()
    if claims.get("type") != GOOGLE_SIGNUP_TOKEN_TYPE:
        raise GoogleAuthError("Invalid or expired Google signup session")

    google_id = claims.get("google_id")
    email = (claims.get("email") or "").strip().lower()
    first_name = (claims.get("first_name") or "").strip()
    last_name = (claims.get("last_name") or "").strip()

    if not google_id or not email or not first_name or not last_name:
        raise GoogleAuthError("Invalid or expired Google signup session")

    return {
        "google_id": str(google_id),
        "email": email,
        "first_name": first_name[:80],
        "last_name": last_name[:80],
    }


def authenticate_google_user(profile: dict) -> tuple[User | None, dict | None]:
    """Return (user, None) on login or (None, pending_signup_dict) for new customers."""
    google_user = User.query.filter_by(google_id=profile["google_id"]).first()
    if google_user:
        return google_user, None

    email_user = User.query.filter_by(email=profile["email"]).first()
    if email_user:
        if email_user.role != "customer":
            raise GoogleAuthError(
                "This email is registered for staff access. Sign in with your password instead."
            )
        if not email_user.is_active:
            raise GoogleAuthError("This account has been deactivated")
        if email_user.must_set_password:
            raise GoogleAuthError(
                "Please use the invite link in your email to set your password before signing in."
            )
        email_user.google_id = profile["google_id"]
        db.session.commit()
        return email_user, None

    return None, {
        "needs_profile": True,
        "signup_token": create_google_signup_token(profile),
        "email": profile["email"],
        "first_name": profile["first_name"],
        "last_name": profile["last_name"],
    }


def complete_google_signup(data: dict) -> tuple[User, dict]:
    profile = _read_google_signup_claims()

    if User.query.filter_by(google_id=profile["google_id"]).first():
        raise GoogleAuthError("A Google account with this profile already exists")
    if User.query.filter_by(email=profile["email"]).first():
        raise GoogleAuthError("An account with this email already exists")

    parish = (data.get("parish") or "").strip()
    if parish not in JAMAICA_PARISHES:
        raise GoogleAuthError("Invalid parish")

    if not data.get("accept_terms"):
        raise GoogleAuthError("You must accept the Terms and Conditions to create an account")

    try:
        contact_number = normalize_phone(data.get("contact_number") or "")
        trn = normalize_trn(data.get("trn"))
    except ValueError as exc:
        raise GoogleAuthError(str(exc)) from exc

    if trn and User.query.filter_by(trn=trn).first():
        raise GoogleAuthError("An account with this TRN already exists")

    from datetime import datetime

    shipping_id = generate_shipping_id()
    user = User(
        email=profile["email"],
        password_hash=hash_password(secrets.token_urlsafe(32)),
        first_name=profile["first_name"],
        last_name=profile["last_name"],
        contact_number=contact_number,
        parish=parish,
        trn=trn,
        shipping_id=shipping_id,
        google_id=profile["google_id"],
        terms_accepted_at=datetime.utcnow(),
    )
    db.session.add(user)
    db.session.commit()

    shipping_address = build_shipping_address(user.shipping_id)
    try:
        send_welcome_email(user.email, user.first_name, user.shipping_id, shipping_address)
    except (EmailServiceError, NotImplementedError) as exc:
        current_app.logger.warning("Welcome email not sent for %s: %s", user.email, exc)

    return user, shipping_address

File: server/src/services/googleAuthService.ts
import { randomBytes } from 'node:crypto'
import { OAuth2Client } from 'google-auth-library'
import type { JwtPayload } from 'jsonwebtoken'
import { eq } from 'drizzle-orm'
import { JAMAICA_PARISHES } from '../constants.js'
import { config } from '../config.js'
import { db } from '../db/index.js'
import { users, type UserRow } from '../db/schema/index.js'
import { hashPassword, normalizePhone } from './authService.js'
import { EmailServiceError, sendWelcomeEmail } from './emailService.js'
import { generateShippingId } from './shippingIdService.js'
import { createGoogleSignupToken } from './tokenService.js'
import { normalizeTrn } from './trnService.js'
import { buildShippingAddress } from './warehouseService.js'

export class GoogleAuthError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'GoogleAuthError'
  }
}

function googleClientId(): string {
  const clientId = config.googleClientId.trim()
  if (!clientId) throw new GoogleAuthError('Google sign-in is not configured')
  return clientId
}

export async function verifyGoogleCredential(credential: string): Promise<{
  google_id: string
  email: string
  first_name: string
  last_name: string
}> {
  if (!credential?.trim()) throw new GoogleAuthError('Google credential is required')

  const client = new OAuth2Client(googleClientId())
  let payload
  try {
    const ticket = await client.verifyIdToken({ idToken: credential.trim(), audience: googleClientId() })
    payload = ticket.getPayload()
  } catch {
    throw new GoogleAuthError('Invalid Google sign-in token')
  }

  if (!payload) throw new GoogleAuthError('Invalid Google sign-in token')
  if (!['accounts.google.com', 'https://accounts.google.com'].includes(payload.iss ?? '')) {
    throw new GoogleAuthError('Invalid Google token issuer')
  }

  const email = (payload.email ?? '').trim().toLowerCase()
  if (!email) throw new GoogleAuthError('Google account email is required')
  if (!payload.email_verified) throw new GoogleAuthError('Google account email is not verified')

  const googleId = payload.sub
  if (!googleId) throw new GoogleAuthError('Invalid Google account')

  const firstName = (payload.given_name ?? email.split('@')[0] ?? '').trim()
  const lastName = (payload.family_name ?? '').trim() || 'Member'

  return {
    google_id: String(googleId),
    email,
    first_name: firstName.slice(0, 80),
    last_name: lastName.slice(0, 80),
  }
}

export function readGoogleSignupClaims(claims: JwtPayload): {
  google_id: string
  email: string
  first_name: string
  last_name: string
} {
  if (claims.type !== 'google_signup') {
    throw new GoogleAuthError('Invalid or expired Google signup session')
  }
  const googleId = claims.google_id as string | undefined
  const email = (claims.email as string | undefined)?.trim().toLowerCase()
  const firstName = (claims.first_name as string | undefined)?.trim()
  const lastName = (claims.last_name as string | undefined)?.trim()
  if (!googleId || !email || !firstName || !lastName) {
    throw new GoogleAuthError('Invalid or expired Google signup session')
  }
  return {
    google_id: String(googleId),
    email,
    first_name: firstName.slice(0, 80),
    last_name: lastName.slice(0, 80),
  }
}

export async function authenticateGoogleUser(profile: {
  google_id: string
  email: string
  first_name: string
  last_name: string
}): Promise<[UserRow | null, Record<string, unknown> | null]> {
  const [googleUser] = await db.select().from(users).where(eq(users.googleId, profile.google_id)).limit(1)
  if (googleUser) return [googleUser, null]

  const [emailUser] = await db.select().from(users).where(eq(users.email, profile.email)).limit(1)
  if (emailUser) {
    if (emailUser.role !== 'customer') {
      throw new GoogleAuthError(
        'This email is registered for staff access. Sign in with your password instead.',
      )
    }
    if (!emailUser.isActive) throw new GoogleAuthError('This account has been deactivated')
    if (emailUser.mustSetPassword) {
      throw new GoogleAuthError(
        'Please use the invite link in your email to set your password before signing in.',
      )
    }
    const [updated] = await db
      .update(users)
      .set({ googleId: profile.google_id })
      .where(eq(users.id, emailUser.id))
      .returning()
    return [updated!, null]
  }

  return [
    null,
    {
      needs_profile: true,
      signup_token: createGoogleSignupToken(profile),
      email: profile.email,
      first_name: profile.first_name,
      last_name: profile.last_name,
    },
  ]
}

export async function completeGoogleSignup(
  claims: JwtPayload,
  data: Record<string, unknown>,
): Promise<[UserRow, ReturnType<typeof buildShippingAddress>]> {
  const profile = readGoogleSignupClaims(claims)

  const [existingGoogle] = await db.select().from(users).where(eq(users.googleId, profile.google_id)).limit(1)
  if (existingGoogle) throw new GoogleAuthError('A Google account with this profile already exists')

  const [existingEmail] = await db.select().from(users).where(eq(users.email, profile.email)).limit(1)
  if (existingEmail) throw new GoogleAuthError('An account with this email already exists')

  const parish = String(data.parish ?? '').trim()
  if (!(JAMAICA_PARISHES as readonly string[]).includes(parish)) {
    throw new GoogleAuthError('Invalid parish')
  }
  if (!data.accept_terms) {
    throw new GoogleAuthError('You must accept the Terms and Conditions to create an account')
  }

  let contactNumber: string
  let trn: string | null
  try {
    contactNumber = normalizePhone(String(data.contact_number ?? ''))
    trn = normalizeTrn(data.trn as string | undefined)
  } catch (err) {
    throw new GoogleAuthError(err instanceof Error ? err.message : String(err))
  }

  if (trn) {
    const [trnUser] = await db.select().from(users).where(eq(users.trn, trn)).limit(1)
    if (trnUser) throw new GoogleAuthError('An account with this TRN already exists')
  }

  const shippingId = await generateShippingId()
  const [user] = await db
    .insert(users)
    .values({
      email: profile.email,
      passwordHash: hashPassword(randomBytes(32).toString('base64url')),
      firstName: profile.first_name,
      lastName: profile.last_name,
      contactNumber,
      parish,
      trn,
      shippingId,
      googleId: profile.google_id,
      termsAcceptedAt: new Date(),
    })
    .returning()

  const shippingAddress = buildShippingAddress(user!.shippingId)
  try {
    sendWelcomeEmail(user!.email, user!.firstName, user!.shippingId, shippingAddress)
  } catch (err) {
    if (err instanceof EmailServiceError) {
      console.warn(`Welcome email not sent for ${user!.email}:`, err)
    }
  }

  return [user!, shippingAddress]
}
